src/python/pants/scm/git.py

```diff
@@ -615,16 +620,17 @@ def _read_object_from_repo(self, rev=None, relpath=None, sha=None):
     header = header.rstrip()
     parts = header.rsplit(SPACE, 2)
     if len(parts) == 2:
-      assert parts[1] == 'missing'
+      assert parts[1] == b'missing'
       raise self.MissingFileException(rev, relpath)
 
     _, object_type, object_len = parts
 
     # Read the object data
     blob = self._cat_file_process.stdout.read(int(object_len))
+    blob = [bytes([b]) for b in bytes(blob)]
```

Comment:
This addition is to consistently use Py3 semantics for iterating through a byte string, which is very different than Py2. See http://python-future.org/compatible_idioms.html#byte-string-literals

Reply:
It's not clear to me why we're doing this... we're reading a precise number of bytes off of the wire, and what we want out of this method is a byte string. So why is the transform necessary?

If the output of the process is currently set as unicode, it shouldn't be.

Reply:
So the reason I did this is to allow us to consistently iterate through the bytestring between Py2 and Py3. 

In Py2, you can do this 
```python
for c in b'test':
  assert type(c) == bytes
```

In Py3, if you try doing this, it will actually return ints, instead of bytechars. If you want to iterate through the byte string like originally intended, you would have to do this
```python
for i in b'test':
  assert type(i) == int
  bytechar = bytes([i])
  assert type(bytechar) == bytes
```

According to the [Future cheat sheet](http://python-future.org/compatible_idioms.html#byte-string-literals), to ensure consistent behavior, you're supposed to wrap the string literal like this
```python
for i in bytes(b'test'):
  bytechar = bytes([i])
```

---

I originally had the above line where it was directly called, e.g. in `_read_tree()`. I moved it here to reduce duplication.

But I think you're right - this should still return a byte string, not a list of byte chars. I'll restore it to returning a byte string, while wrapping it in `bytes()` to get the same semantics in both Py2 and Py3.


P.S. Thanks again for the review! 100% of these changes in this new project phase are manual, so are much harder to reason about, and I appreciate you looking over this all!